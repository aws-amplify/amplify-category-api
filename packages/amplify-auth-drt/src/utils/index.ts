export * from './internal-representation';
export * from './test-utils';
