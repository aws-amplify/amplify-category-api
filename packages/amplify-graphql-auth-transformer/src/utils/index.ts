import { DirectiveWrapper, InvalidDirectiveError } from '@aws-amplify/graphql-transformer-core';
import {
  AppSyncAuthMode,
  TransformerBeforeStepContextProvider,
  TransformerContextProvider,
} from '@aws-amplify/graphql-transformer-interfaces';
import { ObjectTypeDefinitionNode } from 'graphql';
import { Construct } from 'constructs';
import { MODEL_OPERATIONS, READ_MODEL_OPERATIONS } from './constants';
import {
  AuthProvider,
  AuthRule,
  ConfiguredAuthProviders,
  GetAuthRulesOptions,
  ModelOperation,
  RoleDefinition,
  RolesByProvider,
} from './definitions';

export * from './constants';
export * from './definitions';
export * from './validations';
export * from './schema';
export * from './iam';

/**
 * Splits roles into key value pairs by auth type
 */
export const splitRoles = (roles: Array<RoleDefinition>): RolesByProvider => ({
  cognitoStaticRoles: roles.filter((r) => r.static && isAuthProviderEqual(r.provider, 'userPools')),
  cognitoDynamicRoles: roles.filter((r) => !r.static && isAuthProviderEqual(r.provider, 'userPools')),
  oidcStaticRoles: roles.filter((r) => r.static && isAuthProviderEqual(r.provider, 'oidc')),
  oidcDynamicRoles: roles.filter((r) => !r.static && isAuthProviderEqual(r.provider, 'oidc')),
  iamRoles: roles.filter((r) => isAuthProviderEqual(r.provider, 'identityPool')),
  apiKeyRoles: roles.filter((r) => isAuthProviderEqual(r.provider, 'apiKey')),
  lambdaRoles: roles.filter((r) => isAuthProviderEqual(r.provider, 'function')),
});

/**
 * returns @auth directive rules
 */
export const getAuthDirectiveRules = (authDir: DirectiveWrapper, options?: GetAuthRulesOptions): AuthRule[] => {
  const splitReadOperation = (rule: AuthRule): void => {
    const operations: (ModelOperation | 'read')[] = rule.operations ?? [];
    const indexOfRead = operations.indexOf('read', 0);
    if (indexOfRead !== -1) {
      operations.splice(indexOfRead, 1);
      operations.push('get');
      operations.push('list');
      operations.push('listen');
      if (!options?.isSqlDataSource) {
        operations.push('search');
        operations.push('sync');
      }
      // eslint-disable-next-line no-param-reassign
      rule.operations = operations as ModelOperation[];
    }
  };

  const { rules } = authDir.getArguments<{ rules: Array<AuthRule> }>({ rules: [] }, options);
  rules?.forEach((rule) => {
    const operations: (ModelOperation | 'read')[] = rule.operations ?? MODEL_OPERATIONS;

    // In case a customer defines a single dynamic group as a string, put it to an array
    if (rule.groups && typeof rule.groups === 'string') {
      // eslint-disable-next-line no-param-reassign
      rule.groups = [rule.groups];
    }

    if (
      options?.isField &&
      rule.operations &&
      rule.operations.some((operation: ModelOperation | 'read') => operation !== 'read' && READ_MODEL_OPERATIONS.includes(operation))
    ) {
      const offendingOperation = operations.filter((operation) => operation !== 'read' && READ_MODEL_OPERATIONS.includes(operation));
      throw new InvalidDirectiveError(`'${offendingOperation}' operation is not allowed at the field level.`);
    }

    if (operations.includes('read') && operations.some((operation) => operation !== 'read' && READ_MODEL_OPERATIONS.includes(operation))) {
      const offendingOperation = operations.filter((operation) => operation !== 'read' && READ_MODEL_OPERATIONS.includes(operation));
      throw new InvalidDirectiveError(
        `'${offendingOperation}' operations are specified in addition to 'read'. Either remove 'read' to limit access only to '${offendingOperation}' or only keep 'read' to grant all ${READ_MODEL_OPERATIONS} access.`,
      );
    }

    if (!rule.provider) {
      switch (rule.allow) {
        case 'owner':
        case 'groups':
          // eslint-disable-next-line no-param-reassign
          rule.provider = 'userPools';
          break;
        case 'private':
          // eslint-disable-next-line no-param-reassign
          rule.provider = 'userPools';
          break;
        case 'public':
          // eslint-disable-next-line no-param-reassign
          rule.provider = 'apiKey';
          break;
        case 'custom':
          // eslint-disable-next-line no-param-reassign
          rule.provider = 'function';
          break;
        default:
          throw new Error(`Need to specify an allow to assigned a provider: ${rule}`);
      }
    }

    if (isAuthProviderEqual(rule.provider, 'identityPool')) {
      // eslint-disable-next-line no-param-reassign
      rule.generateIAMPolicy = true;
    }

    splitReadOperation(rule);
  });

  return rules;
};

/**
 * gets stack name if the field is paired with function, predictions, or by itself
 */
export const getScopeForField = (
  ctx: TransformerContextProvider,
  obj: ObjectTypeDefinitionNode,
  fieldName: string,
  hasModelDirective: boolean,
): Construct => {
  const fieldNode = obj.fields.find((f) => f.name.value === fieldName);
  const fieldDirectives = fieldNode.directives.map((d) => d.name.value);
  if (fieldDirectives.includes('function')) {
    return ctx.stackManager.getStack('FunctionDirectiveStack');
  }
  if (fieldDirectives.includes('predictions')) {
    return ctx.stackManager.getStack('PredictionsDirectiveStack');
  }
  if (fieldDirectives.includes('sql')) {
    return ctx.stackManager.getStack('CustomSQLStack');
  }
  if (hasModelDirective) {
    return ctx.stackManager.getStack(obj.name.value);
  }
  return ctx.stackManager.scope;
};

/**
 * Returns auth provider passed on config
 */
export const getConfiguredAuthProviders = (context: TransformerBeforeStepContextProvider): ConfiguredAuthProviders => {
  const { authConfig, synthParameters } = context;
  const { adminRoles, identityPoolId } = synthParameters;
  const providers = [
    authConfig.defaultAuthentication.authenticationType,
    ...authConfig.additionalAuthenticationProviders.map((p) => p.authenticationType),
  ];
  const getAuthProvider = (authType: AppSyncAuthMode): AuthProvider => {
    switch (authType) {
      case 'AMAZON_COGNITO_USER_POOLS':
        return 'userPools';
      case 'API_KEY':
        return 'apiKey';
      case 'AWS_IAM':
        return 'identityPool';
      case 'OPENID_CONNECT':
        return 'oidc';
      case 'AWS_LAMBDA':
        return 'function';
      default:
        return 'apiKey';
    }
  };
  const hasIAM = providers.some((p) => p === 'AWS_IAM');
  const hasAdminRolesEnabled = hasIAM && adminRoles?.length > 0;
  /**
   * When AdminUI or generic IAM access is enabled, all the types and operations get IAM auth. If the default auth mode is
   * not IAM all the fields will need to have the default auth mode directive to ensure both IAM and default
   * auth modes are allowed to access
   * default auth provider needs to be added if AdminUI or generic IAM access is enabled and default auth type is not IAM
   */
  const shouldAddDefaultServiceDirective =
    (hasAdminRolesEnabled || context.synthParameters.enableIamAccess) && authConfig.defaultAuthentication.authenticationType !== 'AWS_IAM';
  const configuredProviders: ConfiguredAuthProviders = {
    default: getAuthProvider(authConfig.defaultAuthentication.authenticationType),
    onlyDefaultAuthProviderConfigured: authConfig.additionalAuthenticationProviders.length === 0,
    hasAdminRolesEnabled,
    hasIdentityPoolId: identityPoolId !== null && identityPoolId !== undefined,
    hasApiKey: providers.some((p) => p === 'API_KEY'),
    hasUserPools: providers.some((p) => p === 'AMAZON_COGNITO_USER_POOLS'),
    hasOIDC: providers.some((p) => p === 'OPENID_CONNECT'),
    hasLambda: providers.some((p) => p === 'AWS_LAMBDA'),
    hasIAM,
    shouldAddDefaultServiceDirective,
    genericIamAccessEnabled: synthParameters.enableIamAccess,
  };
  return configuredProviders;
};

export const isAuthProviderEqual = (provider: AuthProvider, otherProvider: AuthProvider): boolean => {
  if (provider === otherProvider) {
    return true;
  }

  if ((provider === 'iam' || provider === 'identityPool') && (otherProvider === 'iam' || otherProvider === 'identityPool')) {
    return true;
  }

  return false;
};
