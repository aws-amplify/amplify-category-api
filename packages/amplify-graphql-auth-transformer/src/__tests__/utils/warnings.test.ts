import { ModelTransformer } from '@aws-amplify/graphql-model-transformer';
import { testTransform } from '@aws-amplify/graphql-transformer-test-utils';
import { TransformerLog, TransformerLogLevel, TransformParameters } from '@aws-amplify/graphql-transformer-interfaces';
import { AuthTransformer } from '../../graphql-auth-transformer';
import { defaultIdentityClaimWarning } from '../../utils/warnings';

describe('defaultIdentityClaimWarning', () => {
  describe('owner based @auth', () => {
    describe('feature flag enabled w/o custom identity claim', () => {
      test('does not return message', () => {
        const context: any = {
          transformParameters: {
            useSubUsernameForDefaultIdentityClaim: true,
            populateOwnerFieldForStaticGroupAuth: true,
          } as TransformParameters,
        };

        expect(defaultIdentityClaimWarning(context, [{ allow: 'owner' }])).toBeUndefined();
      });
    });

    describe('feature flag enabled w/ custom identity claim', () => {
      describe('with default cognito identity claim', () => {
        test('does not return message', () => {
          const context: any = {
            transformParameters: {
              useSubUsernameForDefaultIdentityClaim: true,
              populateOwnerFieldForStaticGroupAuth: true,
            } as TransformParameters,
          };
          expect(defaultIdentityClaimWarning(context, [{ allow: 'owner', identityClaim: 'cognito:username' }])).toBeUndefined();
        });
      });

      describe('with default identity claim', () => {
        test('does not return message', () => {
          const context: any = {
            transformParameters: {
              useSubUsernameForDefaultIdentityClaim: true,
              populateOwnerFieldForStaticGroupAuth: true,
            } as TransformParameters,
          };
          expect(defaultIdentityClaimWarning(context, [{ allow: 'owner', identityClaim: 'username' }])).toBeUndefined();
        });
      });
    });

    describe('feature flag disabled w/ custom identity claim', () => {
      describe('with default cognito identity claim', () => {
        test('does not return message', () => {
          const context: any = {
            transformParameters: {
              useSubUsernameForDefaultIdentityClaim: false,
              populateOwnerFieldForStaticGroupAuth: false,
            } as TransformParameters,
          };
          expect(defaultIdentityClaimWarning(context, [{ allow: 'owner', identityClaim: 'cognito:username' }])).toBeUndefined();
        });
      });

      describe('with default identity claim', () => {
        test('does not return message', () => {
          const context: any = {
            transformParameters: {
              useSubUsernameForDefaultIdentityClaim: true,
              populateOwnerFieldForStaticGroupAuth: true,
            } as TransformParameters,
          };
          expect(defaultIdentityClaimWarning(context, [{ allow: 'owner', identityClaim: 'username' }])).toBeUndefined();
        });
      });
    });

    describe('feature flag disabled w/o custom identity claim', () => {
      test('does return message', () => {
        const context: any = {
          transformParameters: {
            useSubUsernameForDefaultIdentityClaim: false,
            populateOwnerFieldForStaticGroupAuth: false,
          } as TransformParameters,
        };
        expect(defaultIdentityClaimWarning(context, [{ allow: 'owner' }])).toEqual(
          " WARNING: Amplify CLI will change the default identity claim from 'username' " +
            "to use 'sub::username'. To continue using only usernames, set 'identityClaim: \"username\"' on your " +
            "'owner' rules on your schema. The default will be officially switched with v9.0.0. To read " +
            'more: https://docs.amplify.aws/cli/migration/identity-claim-changes/',
        );
      });
    });
  });
});

describe('ownerCanReassignWarning', () => {
  const OWNER_ENABLED_PROVIDERS = ['userPools', 'oidc'];
  const executeTransformAndReturnLogs = (schema: string): string[] =>
    testTransform({
      schema,
      authConfig: {
        defaultAuthentication: { authenticationType: 'API_KEY' },
        additionalAuthenticationProviders: [
          { authenticationType: 'AMAZON_COGNITO_USER_POOLS' },
          {
            authenticationType: 'OPENID_CONNECT',
            openIDConnectConfig: {
              name: 'myOIDCProvider',
              issuerUrl: 'https://some-oidc-provider/auth',
              clientId: 'my-sample-client-id',
            },
          },
        ],
      },
      transformers: [new ModelTransformer(), new AuthTransformer()],
      transformParameters: {
        useSubUsernameForDefaultIdentityClaim: false,
      },
    }).logs;

  beforeEach(() => {
    jest.clearAllMocks();
  });

  OWNER_ENABLED_PROVIDERS.forEach((provider: string) => {
    describe(`${provider} provider`, () => {
      test('warns on owner auth rule without field-level auth', () => {
        expect(
          executeTransformAndReturnLogs(`
          type Blog @model @auth(rules: [{ allow: owner, provider: ${provider} }]) {
            id: ID!
            owner: String
            description: String
          }
        `),
        ).toMatchSnapshot();
      });

      test('does not warn on owner auth rule with field-level auth', () => {
        expect(
          executeTransformAndReturnLogs(`
          type Blog @model @auth(rules: [{ allow: owner, provider: ${provider} }]) {
            id: ID!
            owner: String @auth(rules: [{ allow: owner, provider: ${provider}, operations: [read, delete] }])
            description: String
          }
        `),
        ).toMatchSnapshot();
      });

      test('warns on multiple schemas with multiple reassignable owners each', () => {
        expect(
          executeTransformAndReturnLogs(`
          type Todo @model @auth(rules: [
            { allow: owner, provider: ${provider} }
            { allow: owner, provider: ${provider}, ownerField: "writer" }
            { allow: owner, provider: ${provider}, ownerField: "editors" }
          ]) {
            id: ID!
            description: String
            writer: String
            editors: [String]
            owner: String @auth(rules: [{ allow: owner, provider: ${provider}, operations: [read] }])
          }
          
          type Blog @model @auth(rules: [{ allow: owner, provider: ${provider} }]) {
            id: ID!
            owner: String
            description: String
          }
        `),
        ).toMatchSnapshot();
      });

      test('does not warn on custom owner fields with field-level overrides', () => {
        expect(
          executeTransformAndReturnLogs(`
          type Todo @model @auth(rules: [
            { allow: owner, provider: ${provider} }
            { allow: owner, provider: ${provider} ownerField: "writer" }
            { allow: owner, provider: ${provider} ownerField: "editors" }
          ]) {
            id: ID!
            description: String
            writer: String @auth(rules: [{ allow: owner, provider: ${provider}, operations: [read] }])
            editors: [String] @auth(rules: [{ allow: owner, provider: ${provider}, operations: [read] }])
            owner: String @auth(rules: [{ allow: owner, provider: ${provider}, operations: [read] }])
          }
        `),
        ).toMatchSnapshot();
      });

      test('does not warn on single custom owner fields with field-level override', () => {
        expect(
          executeTransformAndReturnLogs(`
          type Todo @model @auth(rules: [
            { allow: owner, provider: ${provider} ownerField: "writer" }
          ]) {
            id: ID!
            description: String
            writer: String @auth(rules: [{ allow: owner, provider: ${provider}, ownerField: "writer", operations: [read] }])
          }
        `),
        ).toMatchSnapshot();
      });

      test('malformed field-level auth will continue to warn', () => {
        expect(
          executeTransformAndReturnLogs(`
          type Todo @model(subscriptions: null) @auth(rules: [
            { allow: owner, provider: ${provider}, ownerField: "writer" }
            { allow: owner, provider: ${provider}, operations: [read] }
          ]) {
            id: ID!
            description: String
            writer: String @auth(rules: [{ allow: owner, provider: ${provider}, operations: [read] }])
          }
        `),
        ).toMatchSnapshot();
      });

      test('should warn on implicit owner field', () => {
        expect(
          executeTransformAndReturnLogs(`
          type Blog @model @auth(rules: [{ allow: owner, provider: ${provider} }]) {
            id: ID!
            description: String
          }
        `),
        ).toMatchSnapshot();
      });
    });
  });
});

describe('ownerFieldCaseWarning', () => {
  const executeTransformAndReturnLogs = (schema: string): string[] =>
    testTransform({
      schema,
      authConfig: {
        defaultAuthentication: { authenticationType: 'AMAZON_COGNITO_USER_POOLS' },
        additionalAuthenticationProviders: [],
      },
      transformers: [new ModelTransformer(), new AuthTransformer()],
      transformParameters: {
        useSubUsernameForDefaultIdentityClaim: false,
      },
    }).logs;

  test('does not show message with case matching fields', () => {
    const validSchema = `
type Invoice
  @model
  @auth(
    rules: [
      { allow: owner, ownerField: "storeId", operations: [create, read] }
      { allow: owner, ownerField: "customerId", operations: [create, read] }
      { allow: owner }
    ]
  ) {
  id: ID!
  items: [String]
  storeId: ID!
  customerId: ID!
}
`;
    expect(executeTransformAndReturnLogs(validSchema)).toMatchSnapshot();
  });

  test('does not show message with no auth rules', () => {
    const validSchema = `
type Invoice
  @model {
  id: ID!
  items: [String]
  storeId: ID!
  customerId: ID!
}
`;
    expect(executeTransformAndReturnLogs(validSchema)).toMatchSnapshot();
  });
  test('shows message once with one case mismatch in fields', () => {
    const oneCaseMismatchSchema = `
type Invoice
  @model
  @auth(
    rules: [
      { allow: owner, ownerField: "storeID", operations: [create, read] }
      { allow: owner, ownerField: "customerId", operations: [create, read] }
      { allow: owner }
    ]
  ) {
  id: ID!
  items: [String]
  storeId: ID!
  customerId: ID!
}
`;
    expect(executeTransformAndReturnLogs(oneCaseMismatchSchema)).toMatchSnapshot();
  });

  test('shows message twice with two case mismatch in fields', () => {
    const twoCaseMismatchSchema = `
type Invoice
  @model
  @auth(
    rules: [
      { allow: owner, ownerField: "storeID", operations: [create, read] }
      { allow: owner, ownerField: "customerID", operations: [create, read] }
      { allow: owner }
    ]
  ) {
  id: ID!
  items: [String]
  storeId: ID!
  customerId: ID!
}
`;
    expect(executeTransformAndReturnLogs(twoCaseMismatchSchema)).toMatchSnapshot();
  });
  test('shows message with implicit owner field', () => {
    const twoCaseMismatchSchema = `
type Invoice
  @model
  @auth(
    rules: [
      { allow: owner, ownerField: "storeId", operations: [create, read] }
      { allow: owner, ownerField: "customerId", operations: [create, read] }
      { allow: owner }
    ]
  ) {
  id: ID!
  items: [String]
  storeId: ID!
  customerId: ID!
  Owner: String
}
`;
    expect(executeTransformAndReturnLogs(twoCaseMismatchSchema)).toMatchSnapshot();
  });
});

describe('deprecatedIAMProviderWarning', () => {
  const executeTransformAndReturnLogs = (schema: string): Array<TransformerLog> =>
    testTransform({
      schema,
      authConfig: {
        defaultAuthentication: { authenticationType: 'AWS_IAM' },
        additionalAuthenticationProviders: [],
      },
      transformers: [new ModelTransformer(), new AuthTransformer()],
      transformParameters: {
        useSubUsernameForDefaultIdentityClaim: false,
      },
    }).logs;

  test('does not show message when identityPool provider is used', () => {
    const schema = `
type Invoice
  @model
  @auth(
    rules: [
      { allow: private, provider: identityPool }
    ]
  ) {
  id: ID!
  items: [String]
  storeId: ID!
  customerId: ID!
}
`;
    const logs = executeTransformAndReturnLogs(schema);
    expect(logs.length).toBe(0);
  });

  test('shows message when iam provider is used', () => {
    const schema = `
type Invoice
  @model
  @auth(
    rules: [
      { allow: private, provider: iam }
    ]
  ) {
  id: ID!
  items: [String]
  storeId: ID!
  customerId: ID!
}
`;
    const logs = executeTransformAndReturnLogs(schema);
    expect(logs.length).toBe(1);
    const log = logs[0];
    expect(log.level).toBe(TransformerLogLevel.WARN);
    expect(log.message).toBe(
      "WARNING: Schema is using an @auth directive with deprecated provider 'iam'. Replace 'iam' provider with 'identityPool' provider.",
    );
  });
});
