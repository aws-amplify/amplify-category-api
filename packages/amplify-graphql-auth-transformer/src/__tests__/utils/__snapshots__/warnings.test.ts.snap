// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ownerCanReassignWarning oidc provider does not warn on custom owner fields with field-level overrides 1`] = `
Array [
  Object {
    "level": "WARN",
    "message": " WARNING: Amplify CLI will change the default identity claim from 'username' to use 'sub::username'. To continue using only usernames, set 'identityClaim: \\"username\\"' on your 'owner' rules on your schema. The default will be officially switched with v9.0.0. To read more: https://docs.amplify.aws/cli/migration/identity-claim-changes/",
  },
]
`;

exports[`ownerCanReassignWarning oidc provider does not warn on owner auth rule with field-level auth 1`] = `
Array [
  Object {
    "level": "WARN",
    "message": " WARNING: Amplify CLI will change the default identity claim from 'username' to use 'sub::username'. To continue using only usernames, set 'identityClaim: \\"username\\"' on your 'owner' rules on your schema. The default will be officially switched with v9.0.0. To read more: https://docs.amplify.aws/cli/migration/identity-claim-changes/",
  },
]
`;

exports[`ownerCanReassignWarning oidc provider does not warn on single custom owner fields with field-level override 1`] = `
Array [
  Object {
    "level": "WARN",
    "message": " WARNING: Amplify CLI will change the default identity claim from 'username' to use 'sub::username'. To continue using only usernames, set 'identityClaim: \\"username\\"' on your 'owner' rules on your schema. The default will be officially switched with v9.0.0. To read more: https://docs.amplify.aws/cli/migration/identity-claim-changes/",
  },
]
`;

exports[`ownerCanReassignWarning oidc provider malformed field-level auth will continue to warn 1`] = `
Array [
  Object {
    "level": "WARN",
    "message": " WARNING: Amplify CLI will change the default identity claim from 'username' to use 'sub::username'. To continue using only usernames, set 'identityClaim: \\"username\\"' on your 'owner' rules on your schema. The default will be officially switched with v9.0.0. To read more: https://docs.amplify.aws/cli/migration/identity-claim-changes/",
  },
  Object {
    "level": "WARN",
    "message": "WARNING: owners may reassign ownership for the following model(s) and role(s): Todo: [owner]. If this is not intentional, you may want to apply field-level authorization rules to these fields. To read more: https://docs.amplify.aws/cli/graphql/authorization-rules/#per-user--owner-based-data-access.",
  },
]
`;

exports[`ownerCanReassignWarning oidc provider should warn on implicit owner field 1`] = `
Array [
  Object {
    "level": "WARN",
    "message": " WARNING: Amplify CLI will change the default identity claim from 'username' to use 'sub::username'. To continue using only usernames, set 'identityClaim: \\"username\\"' on your 'owner' rules on your schema. The default will be officially switched with v9.0.0. To read more: https://docs.amplify.aws/cli/migration/identity-claim-changes/",
  },
  Object {
    "level": "WARN",
    "message": "WARNING: owners may reassign ownership for the following model(s) and role(s): Blog: [owner]. If this is not intentional, you may want to apply field-level authorization rules to these fields. To read more: https://docs.amplify.aws/cli/graphql/authorization-rules/#per-user--owner-based-data-access.",
  },
]
`;

exports[`ownerCanReassignWarning oidc provider warns on multiple schemas with multiple reassignable owners each 1`] = `
Array [
  Object {
    "level": "WARN",
    "message": " WARNING: Amplify CLI will change the default identity claim from 'username' to use 'sub::username'. To continue using only usernames, set 'identityClaim: \\"username\\"' on your 'owner' rules on your schema. The default will be officially switched with v9.0.0. To read more: https://docs.amplify.aws/cli/migration/identity-claim-changes/",
  },
  Object {
    "level": "WARN",
    "message": "WARNING: owners may reassign ownership for the following model(s) and role(s): Todo: [writer, editors], Blog: [owner]. If this is not intentional, you may want to apply field-level authorization rules to these fields. To read more: https://docs.amplify.aws/cli/graphql/authorization-rules/#per-user--owner-based-data-access.",
  },
]
`;

exports[`ownerCanReassignWarning oidc provider warns on owner auth rule without field-level auth 1`] = `
Array [
  Object {
    "level": "WARN",
    "message": " WARNING: Amplify CLI will change the default identity claim from 'username' to use 'sub::username'. To continue using only usernames, set 'identityClaim: \\"username\\"' on your 'owner' rules on your schema. The default will be officially switched with v9.0.0. To read more: https://docs.amplify.aws/cli/migration/identity-claim-changes/",
  },
  Object {
    "level": "WARN",
    "message": "WARNING: owners may reassign ownership for the following model(s) and role(s): Blog: [owner]. If this is not intentional, you may want to apply field-level authorization rules to these fields. To read more: https://docs.amplify.aws/cli/graphql/authorization-rules/#per-user--owner-based-data-access.",
  },
]
`;

exports[`ownerCanReassignWarning userPools provider does not warn on custom owner fields with field-level overrides 1`] = `
Array [
  Object {
    "level": "WARN",
    "message": " WARNING: Amplify CLI will change the default identity claim from 'username' to use 'sub::username'. To continue using only usernames, set 'identityClaim: \\"username\\"' on your 'owner' rules on your schema. The default will be officially switched with v9.0.0. To read more: https://docs.amplify.aws/cli/migration/identity-claim-changes/",
  },
]
`;

exports[`ownerCanReassignWarning userPools provider does not warn on owner auth rule with field-level auth 1`] = `
Array [
  Object {
    "level": "WARN",
    "message": " WARNING: Amplify CLI will change the default identity claim from 'username' to use 'sub::username'. To continue using only usernames, set 'identityClaim: \\"username\\"' on your 'owner' rules on your schema. The default will be officially switched with v9.0.0. To read more: https://docs.amplify.aws/cli/migration/identity-claim-changes/",
  },
]
`;

exports[`ownerCanReassignWarning userPools provider does not warn on single custom owner fields with field-level override 1`] = `
Array [
  Object {
    "level": "WARN",
    "message": " WARNING: Amplify CLI will change the default identity claim from 'username' to use 'sub::username'. To continue using only usernames, set 'identityClaim: \\"username\\"' on your 'owner' rules on your schema. The default will be officially switched with v9.0.0. To read more: https://docs.amplify.aws/cli/migration/identity-claim-changes/",
  },
]
`;

exports[`ownerCanReassignWarning userPools provider malformed field-level auth will continue to warn 1`] = `
Array [
  Object {
    "level": "WARN",
    "message": " WARNING: Amplify CLI will change the default identity claim from 'username' to use 'sub::username'. To continue using only usernames, set 'identityClaim: \\"username\\"' on your 'owner' rules on your schema. The default will be officially switched with v9.0.0. To read more: https://docs.amplify.aws/cli/migration/identity-claim-changes/",
  },
  Object {
    "level": "WARN",
    "message": "WARNING: owners may reassign ownership for the following model(s) and role(s): Todo: [owner]. If this is not intentional, you may want to apply field-level authorization rules to these fields. To read more: https://docs.amplify.aws/cli/graphql/authorization-rules/#per-user--owner-based-data-access.",
  },
]
`;

exports[`ownerCanReassignWarning userPools provider should warn on implicit owner field 1`] = `
Array [
  Object {
    "level": "WARN",
    "message": " WARNING: Amplify CLI will change the default identity claim from 'username' to use 'sub::username'. To continue using only usernames, set 'identityClaim: \\"username\\"' on your 'owner' rules on your schema. The default will be officially switched with v9.0.0. To read more: https://docs.amplify.aws/cli/migration/identity-claim-changes/",
  },
  Object {
    "level": "WARN",
    "message": "WARNING: owners may reassign ownership for the following model(s) and role(s): Blog: [owner]. If this is not intentional, you may want to apply field-level authorization rules to these fields. To read more: https://docs.amplify.aws/cli/graphql/authorization-rules/#per-user--owner-based-data-access.",
  },
]
`;

exports[`ownerCanReassignWarning userPools provider warns on multiple schemas with multiple reassignable owners each 1`] = `
Array [
  Object {
    "level": "WARN",
    "message": " WARNING: Amplify CLI will change the default identity claim from 'username' to use 'sub::username'. To continue using only usernames, set 'identityClaim: \\"username\\"' on your 'owner' rules on your schema. The default will be officially switched with v9.0.0. To read more: https://docs.amplify.aws/cli/migration/identity-claim-changes/",
  },
  Object {
    "level": "WARN",
    "message": "WARNING: owners may reassign ownership for the following model(s) and role(s): Todo: [writer, editors], Blog: [owner]. If this is not intentional, you may want to apply field-level authorization rules to these fields. To read more: https://docs.amplify.aws/cli/graphql/authorization-rules/#per-user--owner-based-data-access.",
  },
]
`;

exports[`ownerCanReassignWarning userPools provider warns on owner auth rule without field-level auth 1`] = `
Array [
  Object {
    "level": "WARN",
    "message": " WARNING: Amplify CLI will change the default identity claim from 'username' to use 'sub::username'. To continue using only usernames, set 'identityClaim: \\"username\\"' on your 'owner' rules on your schema. The default will be officially switched with v9.0.0. To read more: https://docs.amplify.aws/cli/migration/identity-claim-changes/",
  },
  Object {
    "level": "WARN",
    "message": "WARNING: owners may reassign ownership for the following model(s) and role(s): Blog: [owner]. If this is not intentional, you may want to apply field-level authorization rules to these fields. To read more: https://docs.amplify.aws/cli/graphql/authorization-rules/#per-user--owner-based-data-access.",
  },
]
`;

exports[`ownerFieldCaseWarning does not show message with case matching fields 1`] = `
Array [
  Object {
    "level": "WARN",
    "message": " WARNING: Amplify CLI will change the default identity claim from 'username' to use 'sub::username'. To continue using only usernames, set 'identityClaim: \\"username\\"' on your 'owner' rules on your schema. The default will be officially switched with v9.0.0. To read more: https://docs.amplify.aws/cli/migration/identity-claim-changes/",
  },
  Object {
    "level": "WARN",
    "message": "WARNING: owners may reassign ownership for the following model(s) and role(s): Invoice: [owner]. If this is not intentional, you may want to apply field-level authorization rules to these fields. To read more: https://docs.amplify.aws/cli/graphql/authorization-rules/#per-user--owner-based-data-access.",
  },
]
`;

exports[`ownerFieldCaseWarning does not show message with no auth rules 1`] = `Array []`;

exports[`ownerFieldCaseWarning shows message once with one case mismatch in fields 1`] = `
Array [
  Object {
    "level": "WARN",
    "message": "WARNING: Schema field \\"storeId\\" and ownerField \\"storeID\\" in type Invoice are getting added to your schema but could be referencing the same owner field. If this is not intentional, you may want to change one of the fields to the correct name.
",
  },
  Object {
    "level": "WARN",
    "message": " WARNING: Amplify CLI will change the default identity claim from 'username' to use 'sub::username'. To continue using only usernames, set 'identityClaim: \\"username\\"' on your 'owner' rules on your schema. The default will be officially switched with v9.0.0. To read more: https://docs.amplify.aws/cli/migration/identity-claim-changes/",
  },
  Object {
    "level": "WARN",
    "message": "WARNING: owners may reassign ownership for the following model(s) and role(s): Invoice: [owner]. If this is not intentional, you may want to apply field-level authorization rules to these fields. To read more: https://docs.amplify.aws/cli/graphql/authorization-rules/#per-user--owner-based-data-access.",
  },
]
`;

exports[`ownerFieldCaseWarning shows message twice with two case mismatch in fields 1`] = `
Array [
  Object {
    "level": "WARN",
    "message": "WARNING: Schema field \\"storeId\\" and ownerField \\"storeID\\" in type Invoice are getting added to your schema but could be referencing the same owner field. If this is not intentional, you may want to change one of the fields to the correct name.
",
  },
  Object {
    "level": "WARN",
    "message": "WARNING: Schema field \\"customerId\\" and ownerField \\"customerID\\" in type Invoice are getting added to your schema but could be referencing the same owner field. If this is not intentional, you may want to change one of the fields to the correct name.
",
  },
  Object {
    "level": "WARN",
    "message": " WARNING: Amplify CLI will change the default identity claim from 'username' to use 'sub::username'. To continue using only usernames, set 'identityClaim: \\"username\\"' on your 'owner' rules on your schema. The default will be officially switched with v9.0.0. To read more: https://docs.amplify.aws/cli/migration/identity-claim-changes/",
  },
  Object {
    "level": "WARN",
    "message": "WARNING: owners may reassign ownership for the following model(s) and role(s): Invoice: [owner]. If this is not intentional, you may want to apply field-level authorization rules to these fields. To read more: https://docs.amplify.aws/cli/graphql/authorization-rules/#per-user--owner-based-data-access.",
  },
]
`;

exports[`ownerFieldCaseWarning shows message with implicit owner field 1`] = `
Array [
  Object {
    "level": "WARN",
    "message": "WARNING: Schema field \\"Owner\\" and ownerField \\"owner\\" in type Invoice are getting added to your schema but could be referencing the same owner field. If this is not intentional, you may want to change one of the fields to the correct name.
",
  },
  Object {
    "level": "WARN",
    "message": " WARNING: Amplify CLI will change the default identity claim from 'username' to use 'sub::username'. To continue using only usernames, set 'identityClaim: \\"username\\"' on your 'owner' rules on your schema. The default will be officially switched with v9.0.0. To read more: https://docs.amplify.aws/cli/migration/identity-claim-changes/",
  },
  Object {
    "level": "WARN",
    "message": "WARNING: owners may reassign ownership for the following model(s) and role(s): Invoice: [owner]. If this is not intentional, you may want to apply field-level authorization rules to these fields. To read more: https://docs.amplify.aws/cli/graphql/authorization-rules/#per-user--owner-based-data-access.",
  },
]
`;
